### PR TITLE
HiPE: Various small code cleanups and codegen improvements

### PR DESCRIPTION
**Pull request by margnus1**
The first highlight of this PR is `hipe_rtl: Inline more of match eq test`, which adds a fastpath for match inequality, similar to the `EQ` macro in the VM, except using the slightly more complicated test that neither side is immediate before calling the fallback.
The other highlight is `hipe_beam_to_icode: Omit redtest in some closures` which tweaks the "leafness" criterion that decides whether a function needs to decrease the current process' reductions such that tiny closures that end in a tailcall to a non-closure, such as those generated by `fun F/A`, does not consume any reductions. This removes a lot of argument save/restore bloat from `fun F/A` closures, which become just a single unconditional branch.

### CODE REVIEW

File: lib/hipe/icode/hipe_beam_to_icode.erl [hunk 1]
```diff
@@ -222,25 +222,32 @@ leafness([I|Is], Leafness) ->
 	  end;
 	T when T =:= 'local' orelse T =:= 'remote' ->
 	  {M,F,A} = hipe_icode:enter_fun(I),
-	  case erlang:is_builtin(M, F, A) of
-	    true -> leafness(Is, Leafness);
+	  case erlang:is_builtin(M, F, A)of
```
Comment (kostis) [inline lib/hipe/icode/hipe_beam_to_icode.erl]: Space missing here.

File: lib/hipe/icode/hipe_beam_to_icode.erl [hunk 2]
```diff
@@ -222,25 +222,32 @@ leafness([I|Is], Leafness) ->
...
     selfrec -> true;
+    closure -> false;
     false -> false
   end.
 
 needs_redtest(Leafness) ->
   case Leafness of
     true -> false;
+    %% A "leaf" closure may contain tailcalls to non-closures in addition to
+    %% what othere leaves may contain. Omitting the redtest is safe since they
```
Comment (kostis) [inline lib/hipe/icode/hipe_beam_to_icode.erl]: othere -> other (in this line)
"be" missing in the next one (or something else is wrong in the phrasing)

File: lib/hipe/test/basic_SUITE_data/basic_tuples.erl [hunk 3]
```diff
@@ -73,6 +75,16 @@ get_elements([Element|Rest], Tuple, Pos) ->
 get_elements([], _Tuple, _Pos) ->
   ok.
 
+squares() ->
+  {1*1,   2*2,   3*3,   4*4,   5*5,   6*6,   7*7,   8*8,   9*9,   10*10,
+   11*11, 12*12, 13*13, 14*14, 15*15, 16*16, 17*17, 18*18, 19*19, 20*20}.
+
+test_square() ->
+  %% A bit of gymnastics to create a variable with a 1..19 range
```
Comment (kostis) [inline lib/hipe/test/basic_SUITE_data/basic_tuples.erl]: The comment is a bit confusing -- the next line creates something in the 10..18 range, or not?

More importantly, I do not understand the test here.  All other tests in `test_element/4` call the test and explicitly match for the expected result.  Besides breaking this pattern, what does this one do?

**Comment by sverker** [comment]
The branch has passed our test runs.
Let me know when you are satisfied with the code quality for merge to master.

File: lib/hipe/test/basic_SUITE_data/basic_tuples.erl (hunk 3 above)
Comment (margnus1) [inline lib/hipe/test/basic_SUITE_data/basic_tuples.erl]: The purpose of the test is to exercise the edge case in which the patch causes the `element/2` bounds test to be omitted where it was not before. As for the comment on the variable's range, I was writing about what the range analysis will think (since it cannot assume anything about `rand:uniform/1`). I will rewrite it to be more clear about that as well as explicitly say what edge case it is aiming for.

As for why the test broke the pattern, passing variables with these kinds of ranges tended to widen them to something like `0..256`, which is why I chose the easy solution of putting the entire test in one function.

I will rewrite the test to fit the pattern better, drop the determinism, and be slightly less contrived.

Comment (kostis) [inline lib/hipe/test/basic_SUITE_data/basic_tuples.erl]: If you have noticed that the current range analysis comes up with a `1..19` range in an expression of the form `(E rem 9) + 10`, for some (opaque) `E`, then you have discovered an opportunity to improve the range analysis so that it is able to deduce something better...  Does it really find `1..19` as the range in what you had written there?

Comment (margnus1) [inline lib/hipe/test/basic_SUITE_data/basic_tuples.erl]: You're right. It writes `v6 (-9..9) := 'rem'(v4 (any()), 9) (primop)`, but should be `-8..8`. Further investigation show that `erl_bif_types:arith_rem(neg_inf,pos_inf,9,9) = {-8,8}` produces the right answer, but `hipe_icode_range:range_rem({range,{neg_inf,pos_inf},false},{range,{9,9},false}) = {range,{-9,9},false}` does not. Maybe we should just have `hipe_icode_range` defer to `erl_bif_types` rather than reimplementing everything (plus bugs).

Comment (margnus1) [inline lib/hipe/test/basic_SUITE_data/basic_tuples.erl]: In any case I'll submit that as a separate PR, so feel free to let sverker know when you're satisfied with the code quality of this one.

**Comment by kostis** [comment]
The only thing I do not like is that the test which was added has a dependency on the value of `?SET_LIMIT` which is only mentioned in a comment currently, and is not robust w.r.t. a possible future change of its value to something larger (e.g. 23).  We should look how to improve this test at some point.

Other than that, this PR can be merged now.
